### PR TITLE
tornado 6 compatibility

### LINKED ISSUE #4437: (zmqhandlers.py) RuntimeWarning: coroutine 'WebSocketHandler.get' was never awaited
Attempting to connect to a kernel results in this error being thrown, in a fresh install of the `jupyter` pip package.

Based on [this StackOverflow post](https://stackoverflow.com/a/52824269/8651995) and the fact that [Tornado released a new major version two hours ago](https://pypi.org/project/tornado/#history), I believe this is a compatibility issue with Tornado v6.0

Manually reverting to Tornado v5.1.1 resolved this issue, but Jupyter should be updated or its dependency requirement changed.

### LINKED ISSUE #4439: tornado 6.0 breaks notebook
tornado 6.0 is released at 10AM ET, March 1, 2019, which breaks Jupyter Notebook.
https://pypi.org/project/tornado/6.0/#history

Error message
```
TypeError: 'type' object is not subscriptable
```

How to reproduce & Error trace
```
$ python3 -m venv /home/clouduser/virtualenv 
$ source /home/clouduser/virtualenv/bin/activate
$ pip install pandas jupyter jupyterlab protobuf mysqlclient
$ jupyter notebook --no-browser
Traceback (most recent call last):
  File "/home/clouduser/virtualenv/bin/jupyter-notebook", line 7, in <module>
    from notebook.notebookapp import main
  File "/home/clouduser/virtualenv/lib/python3.5/site-packages/notebook/notebookapp.py", line 47, in <module>
    from zmq.eventloop import ioloop
  File "/home/clouduser/virtualenv/lib/python3.5/site-packages/zmq/eventloop/__init__.py", line 3, in <module>
    from zmq.eventloop.ioloop import IOLoop
  File "/home/clouduser/virtualenv/lib/python3.5/site-packages/zmq/eventloop/ioloop.py", line 21, in <module>
    from tornado import ioloop
  File "/home/clouduser/virtualenv/lib/python3.5/site-packages/tornado/ioloop.py", line 45, in <module>
    from tornado.concurrent import (
  File "/home/clouduser/virtualenv/lib/python3.5/site-packages/tornado/concurrent.py", line 175, in <module>
    future: Union["futures.Future[_T]", "Future[_T]"], value: _T
  File "/usr/lib/python3.5/typing.py", line 552, in __getitem__
    dict(self.__dict__), parameters, _root=True)
  File "/usr/lib/python3.5/typing.py", line 512, in __new__
    for t2 in all_params - {t1} if not isinstance(t2, TypeVar)):
  File "/usr/lib/python3.5/typing.py", line 512, in <genexpr>
    for t2 in all_params - {t1} if not isinstance(t2, TypeVar)):
  File "/usr/lib/python3.5/typing.py", line 190, in __subclasscheck__
    self._eval_type(globalns, localns)
  File "/usr/lib/python3.5/typing.py", line 177, in _eval_type
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

Environments
```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.5 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.5 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial


$python --version
Python 3.5.2

$ pip freeze
attrs==18.2.0
backcall==0.1.0
bleach==3.1.0
decorator==4.3.2
defusedxml==0.5.0
entrypoints==0.3
ipykernel==5.1.0
ipython==7.3.0
ipython-genutils==0.2.0
ipywidgets==7.4.2
jedi==0.13.3
Jinja2==2.10
jsonschema==3.0.1
jupyter==1.0.0
jupyter-client==5.2.4
jupyter-console==6.0.0
jupyter-core==4.4.0
jupyterlab==0.35.4
jupyterlab-server==0.2.0
MarkupSafe==1.1.1
mistune==0.8.4
mysqlclient==1.4.2.post1
nbconvert==5.4.1
nbformat==4.4.0
notebook==5.7.4
numpy==1.16.2
pandas==0.24.1
pandocfilters==1.4.2
parso==0.3.4
pexpect==4.6.0
pickleshare==0.7.5
pkg-resources==0.0.0
prometheus-client==0.6.0
prompt-toolkit==2.0.9
protobuf==3.7.0
ptyprocess==0.6.0
Pygments==2.3.1
pyrsistent==0.14.11
python-dateutil==2.8.0
pytz==2018.9
pyzmq==18.0.0
qtconsole==4.4.3
Send2Trash==1.5.0
six==1.12.0
terminado==0.8.1
testpath==0.4.2
tornado==6.0
traitlets==4.3.2
wcwidth==0.1.7
webencodings==0.5.1
widgetsnbextension==3.4.2
```

On our side, we have to downgrade tornado to 5.1.1 to address the issue.
```
pip install tornado==5.1.1
```

### PR DESCRIPTION
**Pull request by minrk**
to get testing 6.0 again

We usually try to test prereleases to catch this kind of issue, but stopped testing tornado 6 when we found out about a compatibility problem. Whenever we merge a PR to pin dependencies as a band-aid to get tests passing (AOK to get things unblocked), let's try to open a WIP PR unpinning the dependency immediately so we can fix the issue, because it's always a real thing we need to fix, ideally with a release before the upstream package gets its final release.

once this works:

closes #4437
closes #4439

### CODE REVIEW

**Comment by minrk** [comment]
Can release 5.7.5 from 5.7.x after landing this

**Comment by minrk** [comment]
OK, I think this is ready to go and we can release 5.7.5 once it's backported.

**Comment by minrk** [comment]
I think it would be appropriate to pin `tornado<7` in the 5.7.5 release, since we'll likely be making a 6.0 release with bigger changes before long which can be properly unpinned

File: notebook/base/zmqhandlers.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def open(self, *args, **kwargs):
 
     def send_ping(self):
         """send a ping to keep the websocket alive"""
-        if self.stream.closed() and self.ping_callback is not None:
+        if self.ws_connection is None and self.ping_callback is not None:
```
Comment (kevin-bates) [inline notebook/base/zmqhandlers.py]: FWIW - I had found the similar change was necessary via `ws_connection`.

File: notebook/gateway/managers.py [hunk 2]
```diff
@@ -259,9 +258,9 @@ def gateway_request(endpoint, **kwargs):
     except ConnectionRefusedError:
         raise web.HTTPError(503, "Connection refused from Gateway server url '{}'.  " 
               "Check to be sure the Gateway instance is running.".format(GatewayClient.instance().url))
-    except HTTPTimeoutError:
+    except HTTPError:
```
Comment (kevin-bates) [inline notebook/gateway/managers.py]: Thank you.

File: notebook/utils.py [hunk 3]
```diff
@@ -306,3 +324,34 @@ def _check_pid_posix(pid):
...
+# on Python 3
+# TODO: remove monkeypatch after backporting smaller fix to 5.x
+try:
+    import asyncio
+except ImportError:
+    pass
+else:
+    import tornado.gen
+    tornado.gen.maybe_future = maybe_future
+
```
Comment (kevin-bates) [inline notebook/utils.py]: Very nice.  I had also found that `gen.maybe_future` wasn't happy with tornado 6, this is a much more elegant solution for B/C purposes.

**Comment by kevin-bates** [comment]
Thank you Min!  FWIW - I pull your branch and took things for a quick spin against tornado 6.0.1. Looks good! :+1:

**Pull request merged**